### PR TITLE
Nav 140 zipped csv data loader

### PR DESCRIPTION
**Pull request by jsoppela**

### CODE REVIEW

File: navigator_engine/tests/unit_tests/test_data_loaders.py [hunk 1]
```diff
@@ -186,3 +186,29 @@ def test_load_estimates_dataset_without_workflow_state(mock_engine, mocker):
             'data': {'completedTasks': []}
         }
     }
+
+
+def test_load_csv_from_zipped_resource(mock_engine, mocker):
```
Comment (jonathansberry) [inline navigator_engine/tests/unit_tests/test_data_loaders.py:191]: ideally we would have some tests that asserts an empty response is still raised if there is no archive file, or archive file is missing a csv file matching the reg ex.

**Review by jonathansberry — CHANGES_REQUESTED** [review]
Hi Jyri,  Great job starting this.  As you understand more about engine my comments below will likely make more sense.  We are using the data you have loaded to check whether the user can proceed to the next step of their workflow.  If the data legitimately doesn't exist, we should to fail silently here and leave it so that those checks can flag the missing data to the user. If the data is bad/corrupted/wrongly defined then we should raise errors here.

File: navigator_engine/pluggable_logic/data_loaders.py [hunk 2]
```diff
@@ -87,3 +91,29 @@ def load_estimates_dataset(url_key: Hashable, auth_header_key: Hashable, engine:
...
+                if filename_re.match(filename):
+                    matching_filenames.append(filename)
+            if len(matching_filenames) != 1:
+                raise ValueError("Multiple files match filename regex")
+            with zip_file.open(matching_filenames[0], 'r') as csv_file:
+                dataframe = pd.read_csv(csv_file)
+    except zipfile.BadZipFile:
+        raise ValueError('Invalid archive file')
+
+    return dataframe
```
[the remark this answers is not in the document]
Comment (jsoppela) [reply navigator_engine/pluggable_logic/data_loaders.py:119]: Where is `data[resource_type]['url']` defined?

File: navigator_engine/pluggable_logic/data_loaders.py [hunk 3]
```diff
@@ -87,3 +91,67 @@ def load_estimates_dataset(url_key: Hashable, auth_header_key: Hashable, engine:
...
+                    f"Multiple files match filename regex {csv_filename_regex}"
+                    f"in: {data[resource_type]['url']}"
+                )
+            elif len(matching_filenames) == 0:
+                raise ValueError(
+                    f"No files match filename regex {csv_filename_regex}"
+                    f"in: {data[resource_type]['url']}"
+                )
+            with zip_file.open(matching_filenames[0], 'r') as csv_file:
+                dataframe = pd.read_csv(csv_file)
```
Comment (jonathansberry) [inline navigator_engine/pluggable_logic/data_loaders.py:145]:
```suggestion
            elif len(matching_filenames) == 0:
                dataframe = None
            else:
                with zip_file.open(matching_filenames[0], 'r') as csv_file:
                    dataframe = pd.read_csv(csv_file)
```

File: navigator_engine/pluggable_logic/data_loaders.py [hunk 4]
```diff
@@ -87,3 +91,67 @@ def load_estimates_dataset(url_key: Hashable, auth_header_key: Hashable, engine:
...
+            }
+        }
+        return data
+    except ValueError:
+        data[name] = {
+            'data': None,
+            'auth_header': data[resource_type].get('auth_header'),
+            'url': data[resource_type].get('url')
+        }
+        return data
```
Comment (jonathansberry) [inline navigator_engine/pluggable_logic/data_loaders.py:124]: IOError is the wrong error for me to raise.  I will fix this in a later PR. Also, as I mentioned in my last comment I think the assumption should be that data[resource_type] could be empty. Not that an error gets thrown.  I think I need to fix this also in a later PR. Would suggest the following...

```suggestion
    data = load_estimates_dataset_resource(resource_type, auth_header, engine)
    if not data[resource_type]['data']:
        data[name] = {
            'data': None,
            'auth_header': None,
            'url': None
        }
        return data
```

**Review by jonathansberry — CHANGES_REQUESTED** [review]
It was probably my fault for not articulating what I was after clearly enough.  I've just made some suggestions to show precisely what I was thinking.  You should check you understand them and commit them, check the tests pass, before I merge.

File: navigator_engine/tests/unit_tests/test_data_loaders.py [hunk 5]
```diff
@@ -186,3 +186,55 @@ def test_load_estimates_dataset_without_workflow_state(mock_engine, mocker):
...
+
+    with open('navigator_engine/tests/test_data/test_spectrum_file.pjnz', 'rb') as f:
+        spectrum_file = f.read()
+
+    mock_load_estimates_dataset_resource = mocker.patch(
+        'navigator_engine.pluggable_logic.data_loaders.load_estimates_dataset_resource',
+        return_value={
+            'spectrum-file': {
+                'data': spectrum_file,
+                'auth_header': 'Bearer xxxx-xxxx-xxxx-xxxx',
```
Comment (jonathansberry) [inline navigator_engine/tests/unit_tests/test_data_loaders.py:201]: In CKAN's case it is not a bearer token.  I would just get rid of Bearer, although admittedly it's only a cosmetic change for readbility.

File: navigator_engine/tests/unit_tests/test_data_loaders.py [hunk 6]
```diff
@@ -186,3 +186,29 @@ def test_load_estimates_dataset_without_workflow_state(mock_engine, mocker):
...
+
+
+def test_load_csv_from_zipped_resource(mock_engine, mocker):
+
+    with open('navigator_engine/tests/test_data/test_spectrum_file.pjnz', 'rb') as f:
+        spectrum_file = f.read()
+
+    mock_load_estimates_dataset_resource = mocker.patch(
+        'navigator_engine.pluggable_logic.data_loaders.load_estimates_dataset_resource',
+        return_value={'spectrum-file': {'data': spectrum_file}}
```
[the remark this answers is not in the document]
Comment (jonathansberry) [reply navigator_engine/tests/unit_tests/test_data_loaders.py]: This is done :+1:

File: navigator_engine/pluggable_logic/data_loaders.py [hunk 7]
```diff
@@ -87,3 +91,67 @@ def load_estimates_dataset(url_key: Hashable, auth_header_key: Hashable, engine:
...
+                    f"No files match filename regex {csv_filename_regex}"
+                    f"in: {data[resource_type]['url']}"
+                )
+            with zip_file.open(matching_filenames[0], 'r') as csv_file:
+                dataframe = pd.read_csv(csv_file)
+    except zipfile.BadZipFile:
+        raise zipfile.BadZipFile(
+            f"Bad zip file for {resource_type}: {data[resource_type]['url']}"
+        )
+
```
Comment (jonathansberry) [inline navigator_engine/pluggable_logic/data_loaders.py:150]:
```suggestion

    del data[resource_type]
    
```

File: navigator_engine/tests/unit_tests/test_data_loaders.py [hunk 8]
```diff
@@ -186,3 +186,55 @@ def test_load_estimates_dataset_without_workflow_state(mock_engine, mocker):
...
+    assert all(i for i in data.columns == ['Condition checked', 'Status']), \
+        "Unexpected column names in Spectrum check file"
+    assert data.shape == (28, 2), "Unexpected shape of Spectrum check file"
+
+
+def test_load_csv_from_zipped_resource_returns_empty_on_http_error(mock_engine, mocker):
+
+    mocker.patch(
+        'navigator_engine.pluggable_logic.data_loaders.load_estimates_dataset_resource',
+        side_effect=IOError()
```
Comment (jonathansberry) [inline navigator_engine/tests/unit_tests/test_data_loaders.py:228]: Need to mock out empty file not IO Error

**Review by jonathansberry — CHANGES_REQUESTED** [review]
Just some more changes.

**Review by jonathansberry — APPROVED** [review]

**Pull request merged**